src/avram/charms/enum_extensions.cr

```diff
@@ -1,84 +1,53 @@
-macro redeclare_avram_enum(name)
-  {% for member in name.resolve.constants %}
-    {{ member }} = {{ name }}::{{ member }}
-  {% end %}
-end
-
-macro avram_enum(enum_name, &block)
-  {% avram_enum = ("Avram" + enum_name.names.join("::")).id %}
-  enum {{ avram_enum }}
-    {{ block.body }}
+require "./int32_extensions"
 
-    def ===(other : Int32)
-      value == other
-    end
+abstract struct Enum
+  def self.adapter
+    Lucky(self)
   end
 
-  struct {{ enum_name }}
-    def self.adapter
-      Lucky
-    end
-
-    redeclare_avram_enum({{ avram_enum }})
-
-    getter :enum
+  module Lucky(T)
+    include Avram::Type
+    alias ColumnType = Int32
 
-    # You may need to prefix with {{ @type }}
-    #
-    #   {{ @type }}::{{enum_name}}
-    def initialize(@enum : Avram{{ enum_name }})
+    def parse(value : String)
+      if result = T.parse?(value)
+        SuccessfulCast.new(result)
+      else
+        FailedCast.new
+      end
     end
 
-    def initialize(enum_value : Int32)
-      @enum = Avram{{ enum_name }}.from_value(enum_value)
+    def parse(value : Int32)
+      if result = T.from_value?(value)
+        SuccessfulCast.new(result)
+      else
+        FailedCast.new
+      end
     end
 
-    def initialize(enum_value : String)
-      int_value = enum_value.to_i?
-      @enum = if int_value
-                Avram{{ enum_name }}.from_value(int_value)
-              else
-                Avram{{ enum_name }}.parse(enum_value)
-              end
+    def parse(value : T)
+      SuccessfulCast.new(value)
     end
 
-    delegate :===, to_s, to_i, to: @enum
-
-    forward_missing_to @enum
-
-    module Lucky
-      alias ColumnType = Int32
-      include Avram::Type
-
-      def self.criteria(query : T, column) forall T
-        Criteria(T, Int32).new(query, column)
-      end
-
-      def from_db!(value : Int32)
-        {{ enum_name }}.new(value)
-      end
-
-      def parse(value : Avram{{ enum_name }})
-        SuccessfulCast({{ enum_name }}).new(value)
-      end
-
-      def parse(value : String)
-        SuccessfulCast({{ enum_name }}).new({{ enum_name }}.new(value))
-      end
-
-      def parse(value : Int32)
-        SuccessfulCast({{ enum_name }}).new({{ enum_name }}.new(value))
-      end
+    def to_db(value : T)
+      value.value.to_s
+    end
 
-      def to_db(value : Int32)
-        value.to_s
-      end
+    def criteria(query : V, column) forall V
+      Criteria(V, T).new(query, column)
+    end
 
-      def to_db(value : {{ enum_name }})
-        value.value.to_s
+    class Criteria(T, V) < Int32::Lucky::Criteria(T, V)
```

Comment:
Does this mean we restrict Enums to `Int32`? If someone wanted to use `Int16` for example, would this fail?

```crystal
enum Color : Int16
  Red
  Green
end
```

Reply:
Or, actually.. maybe crystal would upcast because Int16 fits inside of Int32 🤔  I guess then what happens with `Int64`? Or do we just not support other ones?

Reply:
To be honest, I had no idea you could change the type like that. This is what we were doing before so I added it back.

Thinking about it, I'm also setting the `Enum::Lucky::ColumnType` to be `Int32` so it's not ok to change the type like that so I think we're fine